tests/playwright-test/expect.spec.ts

```diff
@@ -704,3 +709,155 @@ test('should not leak long expect message strings', async ({ runInlineTest }) =>
+              '\\n\\n' +
+              \`Expected: \${this.utils.printExpected(expected)}\n\` +
+              (matcherResult ? \`Received: \${this.utils.printReceived(matcherResult.actual)}\` : '') +
+              log;
+
+        return {
+          locator,
+          name: 'toHaveAmount',
+          expected,
+          message,
+          pass,
+          actual: matcherResult?.actual,
+          log: matcherResult?.log,
+        };
+      },
+
+      async toBeANicePage(page: Page) {
+        return {
+          name: 'toBeANicePage',
+          expected: 1,
+          message: () => '',
+          pass: true,
+        };
+      }
+    });
+
+    test('custom matchers', async ({ page }) => {
+      await page.setContent(\`
+        <div>
+          <div class='base-amount' data-amount='2'></div>
+        </div>
+      \`);
+      await expect(page.locator('div')).toHaveAmount('3', { timeout: 1000 });
+      await expect(page).toBeANicePage();
+      // @ts-expect-error
+      await expect(page).toHaveAmount('3', { timeout: 1000 });
+      // @ts-expect-error
+      await expect(page.locator('div')).toBeANicePage();
+    });`
+  });
+  expect(result.exitCode).toBe(0);
+});
+
+test('should chain expect matchers and expose matcher utils', async ({ runInlineTest }) => {
+  const result = await runInlineTest({
+    'a.spec.ts': `
+    import { test, expect as baseExpect } from '@playwright/test';
+    import type { Page, Locator } from '@playwright/test';
+
+    function callLogText(log: string[] | undefined): string {
+      if (!log)
+        return '';
+      return log.join('\\n');
+    }
+
+    const expect = baseExpect.extend({
+      async toHaveAmount(locator: Locator, expected: string, options?: { timeout?: number }) {
+        const baseAmount = locator.locator('.base-amount');
+
+        let pass: boolean;
+        let matcherResult: any;
+        try {
+          await baseExpect(baseAmount).toHaveAttribute('data-amount', expected, options);
+          pass = true;
+        } catch (e: any) {
+          matcherResult = e.matcherResult;
+          pass = false;
+        }
+
+        const expectOptions = {
+          isNot: this.isNot,
+        };
+
+        const log = callLogText(matcherResult?.log);
+        const message = pass
+          ? () => this.utils.matcherHint('toBe', locator, expected, expectOptions) +
+              '\\n\\n' +
+              \`Expected: \${this.isNot ? 'not' : ''}\${this.utils.printExpected(expected)}\\n\` +
+              (matcherResult ? \`Received: \${this.utils.printReceived(matcherResult.actual)}\` : '') +
+              log
+          : () =>  this.utils.matcherHint('toBe', locator, expected, expectOptions) +
+              '\\n\\n' +
+              \`Expected: \${this.utils.printExpected(expected)}\n\` +
+              (matcherResult ? \`Received: \${this.utils.printReceived(matcherResult.actual)}\` : '') +
+              log;
+
+        return {
```

Comment:
If I don't return all these fields in my custom matcher, will things keep working?

Reply:
You must return `message` and `pass`. The rest is optional.